beacon-chain/core/helpers/rewards_penalties.go

```diff
@@ -155,6 +156,68 @@ func IncreaseBalance(state state.BeaconState, idx primitives.ValidatorIndex, del
 	return state.UpdateBalancesAtIndex(idx, newBal)
 }
 
+// IncreaseBalanceAndAdjustPrincipalBalance increase validator with the given 'index' balance by 'delta' in Gwei
+// and adjust the principal balance.
+//
+// Spec pseudocode definition:
+//
+// def increase_balance_and_adjust_principal_balance(state: BeaconState, index: ValidatorIndex, delta: Gwei) -> None:
+//
+//	increase_balance(state, index, delta)
+//
+//	validator = state.validators[index]
+//
+//	if state.balances[index] >= validator.principal_balance + delta:
+//
+//		validator.principal_balance += delta
+//
+//	elif state.balances[index] >= validator.principal_balance:
+//
+//		validator.principal_balance = state.balances[index]
+func IncreaseBalanceAndAdjustPrincipalBalance(state state.BeaconState, idx primitives.ValidatorIndex, delta uint64) error {
```

Comment:
Seems rollback is not needed, since returning error will eventually bubble up to ProcessEpoch, which will finally revoke state transition.
If rollback logic is excluded, this function can be much simpler: First `IncreaseBalance` and then update principal balance.

Same idea for decreasing

Reply:
@schyun9212 ^^^^^